Batching of JTAG commands over FTDI for RISCV bulk memory reads/writes

bjoernQ opened the pull request:
This addresses #717 by batching commands instead of doing every operation as request / response

It implements it only for `perform_memory_read_multiple_sysbus` and `perform_memory_write_sysbus` for now to keep the PR small.

There is probably still room for improvements - e.g. all operations could be performed in batches if possible, in future.

However, the results are already encouraging.

e.g `cargo run --features ftdi -- dump 22010000 1000 --chip Riscv` takes
- 6.732151s on _master_
- 253.5766ms on this branch

I was only able to test this with a Sipeed FTDI USB adapter and a BL602 DK

Tiwalun commented:
Thanks for working on this, the speed improvements look great.

We already have batching implemented for the CMSIS-DAP probe, which is used for debugging ARM Cortex-M chips. In that implementation, writes are automatically batched if the probe supports it. There is a `flush` method which can be used to explicitly perform any outstanding writes, or any read will also cause any outstanding writes to be performed, and then return the result of the read.

This has the advantage that there is no separate code for the batching / non-batching case, so the API is simplified. However, it requires a call to `flush` at the appropriate times, otherwise writes might be lost.

You can look at the implementation of `RawDapAccess` for an example:

https://github.com/probe-rs/probe-rs/blob/edfd8bbae96020e629abfb393b8af3de65aaa5f4/probe-rs/src/probe/cmsisdap/mod.rs#L575-L723

I would prefer it if we could keep the RISC-V implementation similar to the ARM implementation. I'm also happy to merge this in the current state, and we can rework the implementation later.

bjoernQ commented:
Thanks for the quick feedback. Seems I totally missed the existing batching implementation.

I really like how the existing implementation is almost transparent to the caller.

Not sure how similar I can make this since the 
> any read will also cause any outstanding writes to be performed, and then return the result of the read
will kill at least the `read_multiple` performance gain but maybe I can come up with something.

My current implementation is really a bit clumsy and I especially dislike the `if self.dtm.supports_batch_access()` in my code.

Will probably take me some time to come up with something here since this is unfortunately just an after-work activity for me.

bjoernQ commented:
I did some changes and finally got rid of the `if self.dtm.supports_batch_access()`

For JLink the batching is now emulated / simulated and that way the same code is used for reading / writing memory.

The schedule function now doesn't directly return the index into the results but returns a wrapper. Also `execute` doesn't return a Vec of CommandResult anymore but a wrapper.

It's not the same as the ARM implementation since that way I don't think there would be any speed improvement for the memory read.

I was only able to test with an FTDI adapter so I tested the JLink changes by copying around the code and tested the batching emulation with JTAG.

Tiwalun reviewed (commented):
I was able to test it with a J-Link, and didn't notice any issues.

I mostly have some small comments, which shouldn't be too much trouble to resolve. 

Error handling is a bit more difficult, because I think it would be nice to not fail completely if any command in a "batch" failed. To resolve this, and maybe even retry commands, is not a trivial task however, so this can also be done at a later point.

bjoernQ commented:
Thanks for the nice and detailed review! Hope I can get to it during the weekend.

I also think the changes to the error handling might be a bit difficult / much effort - will look into it but maybe that needs to be done later.

bjoernQ commented on probe-rs/src/probe/ftdi/mod.rs:
```diff
@@ -565,6 +575,139 @@ impl JTAGAccess for FtdiProbe {
+    }
+
+    fn execute(&mut self) -> Result<Box<dyn CommandResults>, DebugProbeError> {
+        const CHUNK_SIZE: usize = 40;
```
To be honest I did some tests with my setup and I cannot really make sense of this value. The largest I was able to do during testing was 45

I cannot relate this to e.g. the size of the data written or read.

The reported write/read chunk size from libftdi is 4096 both which I don't even come close to with this value.
Maybe I can just comment the line accordingly

bjoernQ commented on probe-rs/src/probe/ftdi/commands.rs:
```diff
@@ -0,0 +1,455 @@
+        let mut data: &[u8] = &self.data;
+        while bits > 0 {
+            if bits >= 8 {
+                command.extend_from_slice(&[0x4b, 0x07, data[0]]);
```
to be honest I just copied over the code from https://github.com/probe-rs/probe-rs/blob/a920c84c8a2942b650e37bdba963bee719543efc/probe-rs/src/probe/ftdi/mod.rs#L94

in my understanding 0x4b is the FTDI command "Clock Data to TMS pin (no read)"

bjoernQ replied:
I added a comment about it

bjoernQ commented on probe-rs/src/probe/ftdi/mod.rs:
```diff
@@ -565,6 +575,139 @@ impl JTAGAccess for FtdiProbe {
+                cmd.add_bytes(&mut out_buffer);
+                size += cmd.bytes_to_read();
+            }
+            out_buffer.push(0x87); // Send Immediate: This will make the chip flush its buffer back to the PC.
```
I'll explain in the comment that this is an FTDI command and where to find details about it

bjoernQ commented on probe-rs/src/probe/ftdi/mod.rs:
```diff
@@ -565,6 +575,139 @@ impl JTAGAccess for FtdiProbe {
+                            None => results.push(data),
+                        }
+                    }
+                    Err(e) => return Err(e),
```
Briefly looking into this I also think it's a big change and probably better to do it later

bjoernQ commented:
I hope I addressed all issues but the error reporting. I think that added some really good improvements.

Looking into the error reporting briefly I totally agree it should and could be improved in future but as you said it's going to be more effort.

The thing about not being able to reasonable explain the chunk size really nags me - maybe someone else can explain it - I tried to point out the weirdness about it in the comment.

Should I open an issue about the need to refine the error reporting or should I add a comment there to not forget about it?

Tiwalun reviewed (approved):
Looks good, thanks for implementing this!

Are you planning to do more work on this, to update other RISCV functions to the new API as well?

bors r+

bjoernQ commented:
Great and thank you! I'd love to port more functionality to this new API but cannot promise anything regarding when I get to it. Probably the first thing I have to do is looking into the error handling improvements.

There are also some additional things I'd like to look into e.g.
- how should CSR registers get exposed to the user
- how could we implement custom reset sequences like https://github.com/riscv/riscv-openocd/blob/riscv/tcl/target/gd32vf103.cfg#L39-L85 or https://github.com/bouffalolab/bl_iot_sdk/blob/master/tools/debug/tgt_602.cfg#L17-L22

Tiwalun commented:
Don't know if you've seen it yet, but we're currently working on sequences for the Cortex-M chips, in #483. RISCV sequences could probably be implemented in a similar way. I've also run into problems with the GD32VF103 because of missing sequences, life would be easier if people would just implement the spec :roll_eyes:

bjoernQ commented:
Thanks for the pointer to #483 - totally missed that

I actually like the "hardcoded" approach much more than the config-file-hell with OpenOCD - if the chip is supported it will just work this way.

I guess that for RV32 there are even more "special-snowflakes" that need some custom handling - really a pity that some people not just implement the spec

Yatekii commented:
The idea behind sequences is to just have the very minimal bits in those sequences. We want to avoid exactly what's a problem in OpenOCD. Tho, hardcoding every exception is a huge problem as well. You loose every notion of overview soon (it's a problem in OpenOCD too as they hardcode a lot of things as well).
So we try to get the best of both worlds here and push as little as possible to the sequences.

I hope that makes sense :)
